src/utils/router-utils.js

```diff
@@ -59,6 +59,12 @@ export async function scrollBehavior(to, from, savedPosition) {
     const offset = baseNavOffset + apiChangesNavHeight + getExtraScrollOffset(to);
 
     const y = process.env.VUE_APP_TARGET === 'ide' ? 0 : offset;
+
+    // Wait until the page has been rendered if the navigation to a hash is
+    // triggered from a previous route.
+    if (from) {
```

Comment:
@dobromir-hristov, assuming this change looks good to you...do you know _why_ I need this conditional? I would think that adding the `await ...` call would work in all scenarios but it seems to break loading the URL directly for some reason (basically inverting the original problem)

Reply:
We use `scrollBehavior` in basically two places, the router setup and the `mixins/scrollToElement`.

The problem is the router has no idea an async fetching process is running, so it can await it before scrolling to the element. Which is probably why, we added another call via the `mixins/OnPageScrollToFragment`, which once the `DocumentationTopic` is mounted, calls the `scrollToElement` function.

This on its own is wrong, because we now try to scroll to the right part twice.

Then, I saw we have this call inside `mixins/scrollToElement`

```js
      } = await this.$router.options.scrollBehavior(resolvedRoute.route);
```

This probably changes the `this` context of that `scrollBehavior` method, which is why you may not be able to call `await this.$app.nextTick()` inside.

```diff
Index: src/mixins/scrollToElement.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/mixins/scrollToElement.js b/src/mixins/scrollToElement.js
--- a/src/mixins/scrollToElement.js	(revision 48ec05a4c7969aeaef28479a27c9633311a8f90f)
+++ b/src/mixins/scrollToElement.js	(date 1670315962975)
@@ -19,7 +19,7 @@
       const {
         selector,
         offset,
-      } = await this.$router.options.scrollBehavior(resolvedRoute.route);
+      } = await this.$router.options.scrollBehavior.call(this.$router, resolvedRoute.route);
       const element = document.querySelector(selector);
       if (!element) return null;
 
Index: src/utils/router-utils.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/utils/router-utils.js b/src/utils/router-utils.js
--- a/src/utils/router-utils.js	(revision 48ec05a4c7969aeaef28479a27c9633311a8f90f)
+++ b/src/utils/router-utils.js	(date 1670316291651)
@@ -62,9 +62,7 @@
 
     // Wait until the page has been rendered if the navigation to a hash is
     // triggered from a previous route.
-    if (from) {
-      await this.app.$nextTick();
-    }
+    await this.app.$nextTick();
     return { selector: cssEscapeTopicIdHash(hash), offset: { x: 0, y } };
   }
   if (areEquivalentLocations(to, from)) {
```

Reply:
We may do something like this at some time, where we await the data fetch promise - https://stackoverflow.com/questions/74008705/how-to-use-vue-router-scroll-behavior-with-page-transition

Reply:
yep exactly what I said :) Its barely noticeable atm though, so deff something to refine if someone is not super busy.

Reply:
Hm. I spent a lot of time playing with this today and didn't make any meaningful progress :(

I was able to set up [something](http://github.com/apple/swift-docc-render/compare/main...mportiz08:improve-scrolling-to-anchors) quick/dirty to wait on the fetch requests and also the next tick—however, I still had to wait for an arbitrary number of frames for it to work in all scenarios, possibly because of things like images still taking additional time to load.

Reply:
I think after it downloads the JSON, it needs a few moments to renders all the images AND the code listings, as those come in a few moments later.